File: RpnCalculator-Scala/src/rpn/scala/RpnCalculator.scala
package rpn.scala

import operators.AddOperator
import operators.Operator
import operators.SubtractOperator

class RpnCalculator {
  private val numbers = new RpnStack

  def enter(operand: Int): scala.Unit = {
    numbers.push(operand)
  }

  def perform(operatorCharacter: String): Int = {
    val registry = new OperatorRegistry()
    val operator = registry.getOperator(operatorCharacter)
    return operator.performOperation(numbers);
  }
}

File: src/main/java/rpn/RpnCalculator.java
package rpn;

import rpn.operator.Operator;

public class RpnCalculator {
    private RpnStack numbers = new RpnStack();
    private OperatorRegistry operatorRegistry = new OperatorRegistry();

    public void enter(Integer operand) {
        numbers.push( operand );
    }

    public int perform(String operatorCharacter) {
        int result = 0;
        Operator operatorImplementation = operatorRegistry.getOperator( operatorCharacter );
        if (operatorImplementation == null)
            throw new IllegalArgumentException( "Unknown Operator " + operatorCharacter );
        result = operatorImplementation.doOperation( numbers );
        return result;
    }

    public int peek() {
        return numbers.peek();
    }

}
